---
title: Documentation of the DNA meta-mutator
output: 
  html_document:
    code_folding: hide
---


# General info

Goal : mutates a certain gene

Input : teleonomical ID of the progtein to mutate
Output : nothing, only mutated DNA


The meta-mutator calls individual mutators which can mutate:
- a single codon
- a aggregated codon sequence (whole sequence that will be aggregated into a single codon)
- an aggregator
- another mutator
- create a new gene

## mutators

Before describing how the meta-mutator works, we document here what are the different kinds of mutators.


## Description

The DNA meta-mutator first finds the gene corresonding to the progtein whose ID was given in input.
Then it reads sequentially the DNA, detects metagenetic data if there is some, 
in which case it calls the corresponding mutator to mutate the code.


### Step 0 : initialisation


### Step 1 : detecting mutator


### Step 2 : end of DNA adjustment


### Step 3 : outro


# Variables

var 1 : DNA and gene pos in body
var 2 : value at pos in body
var 3 : condition var
var 4 : condition var
var 5 : GSTART_ID
var 6 : func ID to mutate
var 7 : genetic shift

# Pseudo-code

```{pseudo class.source='foldable'}
*find DNA in body*
var1 <- 0
While gene at var1 != MARKER and gene at var1+2 != -1:
  var 1+=7


*find func in DNA
While gene at var1 != GSTART and gene at var1+1 != func ID (var6):
  var 1+=7


*mutate code*
while var1 != GSTART_ID
  get DNA at var1 in var2
  if var2 is a mutator ID (i.e. 0 <= var2 - mutator_id < + 10000)
    put DNA ptr and shift onto the stack
    call mutator (var2)
    get DNA ptr and shift from stack

  increment var1 (pos in DNA)

shift to be a multiple of SIZE_INSTR
-> if var7 equals n, put n zeros at var1

end
```

# Code

## DNA code

### Raw code

```{DNA class.source='foldable'}
GSTART_ID, 107, id_tRNA_CVARS, LV_place, 7, id_tRNA_IVARS, id_tRNA_SLVcst, 
5, GSTART_ID-1, id_tRNA_INCLV, 0, 5, id_tRNA_SP_GSPLV, 6, 
id_tRNA_INCLV, 0, 6, id_tRNA_SLVDLV, 6, 6, id_tRNA_SLVcst, 
1, -7, id_tRNA_SLVcst, 3, 1, id_tRNA_IF0, 3, 
id_tRNA_OPELVcst, 0, 1, 7, id_tRNA_RGLVLV, 2, 1, 
id_tRNA_CMPLVcst, 0, 2, instruction::MARKER, 4, id_tRNA_IF0, 4, 
id_tRNA_OPELVcst, 0, 1, 2, id_tRNA_RGLVLV, 2, 1, 
id_tRNA_CMPLVcst, 0, 2, -1, 3, id_tRNA_INCLV, 1, 
3, id_tRNA_OPELVcst, 1, 1, 2, id_tRNA_SIJ, id_tRNA_SEJ, 
id_tRNA_SWJ, id_tRNA_SLVcst, 3, 1, id_tRNA_IF0, 3, id_tRNA_OPELVcst, 
0, 1, 7, id_tRNA_RGLVLV, 2, 1, id_tRNA_CMPLVLV, 
0, 2, 5, 4, id_tRNA_IF0, 4, id_tRNA_INCLV, 
0, 1, id_tRNA_RGLVLV, 2, 1, id_tRNA_CMPLVLV, 0, 
2, 6, 3, id_tRNA_INCLV, 1, 3, id_tRNA_INCLV, 
1, 1, id_tRNA_SIJ, id_tRNA_SEJ, id_tRNA_SWJ, id_tRNA_INCLV, 0, 
1, id_tRNA_SLVcst, 7, 0, id_tRNA_SLVcst, 3, 1, 
id_tRNA_IF0, 3, id_tRNA_RGLVLV, 2, 1, id_tRNA_OPELVcst, 1, 
2, id_begin_mutators, id_tRNA_CMPLVcst, 2, 2, 0, 3, 
id_tRNA_CMPLVcst, 2, 2, 10000, 4, id_tRNA_OPELVLV, 2, 
3, 4, id_tRNA_IF0, 3, id_tRNA_ScstLV, 99, 1, 
id_tRNA_INCcst, 1, 99, id_tRNA_ScstLV, 99, 7, id_tRNA_CALLLV, 
2, id_tRNA_SP_GSPLV, 1, id_tRNA_SP_GSPLV, 7, id_tRNA_INCLV, 1, 
7, id_tRNA_SLVDLV, 1, 1, id_tRNA_SLVDLV, 7, 7, 
id_tRNA_SIJ, id_tRNA_SEJ, id_tRNA_INCLV, 0, 1, id_tRNA_RGLVLV, 2, 
1, id_tRNA_CMPLVLV, 0, 2, 5, 3, id_tRNA_INCLV, 
1, 3, id_tRNA_SWJ, id_tRNA_IF0, 7, id_tRNA_INCLV, 1, 
7, id_tRNA_GADD, 1, 0, id_tRNA_INCLV, 0, 1, 
id_tRNA_SWJ, id_tRNA_RET, GSTOP_ID, 0, 0, 0, 0, 
```


### Commented code

```{pseudoDNA class.source='foldable'}
*init stuff*
id_tRNA_SLVcst, 5, GSTART_ID-1
id_tRNA_INCLV, 0, 5               // var5 <- GSTART_ID
id_tRNA_SP_GSPLV, 6
id_tRNA_INCLV, 0, 6
id_tRNA_SLVDLV, 6, 6              // var6 <- func ID (1 after TOPS)


*find DNA in body*
id_tRNA_SLVcst, 1, -7             // var1 <- -7
id_tRNA_SLVcst, 3, 1              // var3 <- 1
id_tRNA_IF0, 3
| id_tRNA_OPELVcst, 0, 1, 7       // var1+=7
| id_tRNA_RGLVLV, 2, 1            // var2 <- body val at var1
| id_tRNA_CMPLVcst, 0, 2, instruction::MARKER, 4    // var4 <- var2==MARKER
| id_tRNA_IF0, 4
| | id_tRNA_OPELVcst, 0, 1, 2       // var1+=2
| | id_tRNA_RGLVLV, 2, 1            // var2 <- body val at var1
| | id_tRNA_CMPLVcst, 0, 2, -1, 3   // var3 <- var2==-1
| | id_tRNA_INCLV, 1, 3             // var3--
| | id_tRNA_OPELVcst, 1, 1, 2       // var1-=2
| id_tRNA_SIJ
| id_tRNA_SEJ
id_tRNA_SWJ


*find func in DNA
id_tRNA_SLVcst, 3, 1              // var3 <- 1
id_tRNA_IF0, 3
| id_tRNA_OPELVcst, 0, 1, 7       // var1+=7
| id_tRNA_RGLVLV, 2, 1            // var2 <- body val at var1
| id_tRNA_CMPLVLV, 0, 2, 5, 4     // var4 <- var2==GSTART_ID (var5)
| id_tRNA_IF0, 4
| | id_tRNA_INCLV, 0, 1             // var1++
| | id_tRNA_RGLVLV, 2, 1            // var2 <- body val at var1
| | id_tRNA_CMPLVLV, 0, 2, 6, 3     // var3 <- var2==var6
| | id_tRNA_INCLV, 1, 3             // var3--
| | id_tRNA_INCLV, 1, 1             // var1--
| id_tRNA_SIJ
| id_tRNA_SEJ
id_tRNA_SWJ
id_tRNA_INCLV, 0, 1                 // var1++ (just after GSTART)

*init genetic shift*
id_tRNA_SLVcst, 7, 0                // var7 <- 0

*mutate code*
id_tRNA_SLVcst, 3, 1                // var3 <- 1
id_tRNA_IF0, 3
| id_tRNA_RGLVLV, 2, 1                            // var2 <- body val at var1
| | id_tRNA_OPELVcst, 1, 2, id_begin_mutators     // var2 -= mutator_ids
| | id_tRNA_CMPLVcst, 2, 2, 0, 3                  // var3 <- var 2 >= 0 ?
| | id_tRNA_CMPLVcst, 2, 2, 10000, 4              // var4 <- var 2 < 10000 ?
| | id_tRNA_OPELVLV, 2, 3, 4                      // var3 *= var4 (1 if mut <= var4 < mut+10k)
| id_tRNA_IF0, 3
| |
| | *put DNA ptr and shift onto the stack*
| | id_tRNA_ScstLV, 99, 1               // put var1 at p
| | id_tRNA_INCcst, 1, 99
| | id_tRNA_ScstLV, 99, 7               // put var7 at p-1
| | 
| | *call mutator*
| | id_tRNA_CALLLV, 2                   // call mutator
| | 
| | *get DNA ptr and shift from stack*
| | id_tRNA_SP_GSPLV, 1                 // var1 <- TOPS  (p)
| | id_tRNA_SP_GSPLV, 7                 // var7 <- TOPS  (p)
| | id_tRNA_INCLV, 1, 7                 // var7-- (p-1)
| | id_tRNA_SLVDLV, 1, 1                // var1 <- DNA ptr
| | id_tRNA_SLVDLV, 7, 7                // var7 <- DNA shift
| |
| id_tRNA_SIJ
| id_tRNA_SEJ
|
| id_tRNA_INCLV, 0, 1             // var1++
| id_tRNA_RGLVLV, 2, 1            // var2 <- body val at var1
| id_tRNA_CMPLVLV, 0, 2, 5, 3     // var3 <- var2==GSTART_ID (var5)
| id_tRNA_INCLV, 1, 3             // var3--
id_tRNA_SWJ

*shift to be a multiple of SIZE_INSTR*
id_tRNA_IF0, 7                    // insert 0 until there is no  more shift
| id_tRNA_INCLV, 1, 7             // var7--
| id_tRNA_GADD, 1, 0              // genes[var1] <- 0
| id_tRNA_INCLV, 0, 1             // var1++
id_tRNA_SWJ

id_tRNA_RET
```


## Teleonomical code

### XASM code

```{XASM class.source='foldable'}
instruction::MARKER, 0, 107, 0, 0, 0, 0,

*coded genetically only*

instruction::JMP, 2, 0, 0, 0, 0, 0,

instruction::MARKER, 0, 107, 0, 0, 0, 0,
```

### Raw code

```{Raw class.source='foldable'}
46, 0, 107, 0, 0, 0, 0,
2, 1, 0, 0, 104, -51452, 0,
5, 1, 0, 0, 104, 0, 0,
2, 1, 1, 0, 105, 99, 0,
5, 1, 0, 0, 105, 0, 0,
2, 1, 2, 0, 105, 105, 0,
2, 1, 0, 0, 100, -7, 0,
2, 1, 0, 0, 102, 1, 0,
27, 0, 1, 0, 18, 102, 0,
7, 1, 1, 0, 100, 100, 7,
39, 1, 2, 0, 101, 100, 0,
27, 0, 1, 0, 3, 101, 46,
2, 1, 0, 0, 103, 0, 0,
20, 0, 0, 0, 2, 0, 0,
2, 1, 0, 0, 103, 1, 0,
27, 0, 1, 0, 10, 103, 0,
7, 1, 1, 0, 100, 100, 2,
39, 1, 2, 0, 101, 100, 0,
27, 0, 1, 0, 3, 101, -1,
2, 1, 0, 0, 102, 0, 0,
20, 0, 0, 0, 2, 0, 0,
2, 1, 0, 0, 102, 1, 0,
6, 1, 0, 0, 102, 0, 0,
8, 1, 1, 0, 100, 100, 2,
20, 0, 0, 0, 1, 0, 0,
21, 0, 0, 0, 17, 0, 0,
2, 1, 0, 0, 102, 1, 0,
27, 0, 1, 0, 18, 102, 0,
7, 1, 1, 0, 100, 100, 7,
39, 1, 2, 0, 101, 100, 0,
27, 0, 1, 1, 3, 101, 104,
2, 1, 0, 0, 103, 0, 0,
20, 0, 0, 0, 2, 0, 0,
2, 1, 0, 0, 103, 1, 0,
27, 0, 1, 0, 10, 103, 0,
5, 1, 0, 0, 100, 0, 0,
39, 1, 2, 0, 101, 100, 0,
27, 0, 1, 1, 3, 101, 105,
2, 1, 0, 0, 102, 0, 0,
20, 0, 0, 0, 2, 0, 0,
2, 1, 0, 0, 102, 1, 0,
6, 1, 0, 0, 102, 0, 0,
6, 1, 0, 0, 100, 0, 0,
20, 0, 0, 0, 1, 0, 0,
21, 0, 0, 0, 17, 0, 0,
5, 1, 0, 0, 100, 0, 0,
2, 1, 0, 0, 106, 0, 0,
2, 1, 0, 0, 102, 1, 0,
27, 0, 1, 0, 31, 102, 0,
39, 1, 2, 0, 101, 100, 0,
8, 1, 1, 0, 101, 101, 40000,
29, 0, 1, 0, 3, 101, 0,
2, 1, 0, 0, 102, 0, 0,
20, 0, 0, 0, 2, 0, 0,
2, 1, 0, 0, 102, 1, 0,
29, 0, 1, 0, 3, 101, 10000,
2, 1, 0, 0, 103, 0, 0,
20, 0, 0, 0, 2, 0, 0,
2, 1, 0, 0, 103, 1, 0,
9, 1, 1, 1, 102, 102, 103,
27, 0, 1, 0, 11, 102, 0,
2, 1, 1, 0, 99, 100, 0,
6, 1, 0, 0, 99, 0, 0,
2, 1, 1, 0, 99, 106, 0,
49, 0, 1, 0, 0, 101, 0,
2, 1, 1, 0, 100, 99, 0,
2, 1, 1, 0, 106, 99, 0,
6, 1, 0, 0, 106, 0, 0,
2, 1, 2, 0, 100, 100, 0,
2, 1, 2, 0, 106, 106, 0,
20, 0, 0, 0, 1, 0, 0,
5, 1, 0, 0, 100, 0, 0,
39, 1, 2, 0, 101, 100, 0,
27, 0, 1, 1, 3, 101, 104,
2, 1, 0, 0, 102, 0, 0,
20, 0, 0, 0, 2, 0, 0,
2, 1, 0, 0, 102, 1, 0,
6, 1, 0, 0, 102, 0, 0,
21, 0, 0, 0, 30, 0, 0,
27, 0, 1, 0, 5, 106, 0,
6, 1, 0, 0, 106, 0, 0,
44, 1, 0, 0, 100, 0, 0,
5, 1, 0, 0, 100, 0, 0,
21, 0, 0, 0, 4, 0, 0,
19, 2, 0, 0, 0, 0, 0,
46, 0, 107, 0, 0, 0, 0,
```